internal_packages/keybase/lib/encrypt-button.cjsx

```diff
@@ -0,0 +1,140 @@
+    # periodically check to see if we should get any keys
+    @interval = setInterval(@_onTrigger, 100)
+
+  componentWillUnmount: ->
+    @unlistenKeystore()
+    clearInterval(@interval)
+
+  _getStateFromStores: ->
+    keys = []
+    DraftStore.sessionForClientId(@props.draftClientId).then (session) =>
+      for recipient in session.draft().to
+        publicKeys = PGPKeyStore.pubKeys(recipient.email)
+        if publicKeys.length < 1
+          # no key for this user:
+          # push a null so that @_encrypt can line this array up with the
+          # array of recipients
+          keys.push({address: recipient.email, key: null})
+        else
+          # note: this, by default, encrypts using every public key associated
+          # with the address
+          for publicKey in publicKeys
+            if not publicKey.key?
+              PGPKeyStore.getKeyContents(key: publicKey)
+            else
+              keys.push(publicKey)
+
+    return {
+      keys: keys
+    }
+
+  _onKeystoreChange: =>
+    if @state.currentlyEncrypted
+      # message is no longer encrypted with all keys - decrypt it!
+      @_toggleCrypt()
+    # now actually get the new keys
+    @setState(@_getStateFromStores())
+
+  _onTrigger: =>
+    # Some sort of outside event has occured that we must respond to
+    DraftStore.sessionForClientId(@props.draftClientId).then (session) =>
+      if @state.currentlyEncrypted
+        # THERE CAN BE NO CHANGES
+        if @state.cryptotext? and @state.cryptotext != ""
+          # can't put HTML in or it will be sent with the message, confusing recipients
+          #body = '<div style="background-color: #D3D3D3; padding: 10px;">' + @state.cryptotext + '</div>'
+          body = @state.cryptotext
+          session.changes.add({body: body})
+      else
+        plaintext = session.draft().body
+        @setState({plaintext: plaintext})
+
+    @setState(@_getStateFromStores())
+
+  render: ->
+    <div className="n1-keybase">
+      <button title="Encrypt email body" className="btn btn-toolbar" onClick={ => @_onClick()} ref="button">
+        Encrypt
+      </button>
+    </div>
+
+  _onClick: =>
+    @_toggleCrypt()
+
+  _toggleCrypt: =>
+    # if decrypted, encrypt, and vice versa
+    DraftStore.sessionForClientId(@props.draftClientId).then (session) =>
+      if @state.currentlyEncrypted
+        # if the message is already encrypted, place the stored plaintext back
+        # in the draft (i.e. un-encrypt)
+        session.changes.add({body: @state.plaintext})
+        @setState({currentlyEncrypted: false})
+      else
+        # if not encrypted, save the plaintext, then encrypt
+        plaintext = session.draft().body
+        @_encrypt(plaintext, @state.keys)
+        @setState({currentlyEncrypted: true, plaintext: plaintext})
+
+  _encrypt: (text, keys) =>
+    # addresses which don't have a key
+    nullAddrs = _.pluck(_.filter(@state.keys, (key) -> return key.key is null), "address")
```

Comment:
It seems like we're constantly maintaining a set of keys on `state`. What if we just fetched the keys when you click encrypt instead?